### PR TITLE
Benchmarks that runs server and client and separate processes.

### PR DESCRIPTION
**Pull request by MakMukhi**
Also, 

* Record CPU utilization on both client and server.
* Profile CPU and Memory on both client and server.
* Do not record buckets.
* One stop shell script to run benchmarks.

Sample output:

```
$ ./run_bench.sh -r 1,10 -c 1,200 -req 1,100 -resp 1 -rpc_type unary,streaming
================================================================================
r_1_c_1_req_1_resp_1_unary_1522345524
qps: 9247.9
Latency: (50/90/99 %ile): 102.367µs/122.551µs/168.352µs
Client CPU utilization: 15.369409788s
Client CPU profile: /tmp/client_r_1_c_1_req_1_resp_1_unary_1522345524.cpu
Client Mem Profile: /tmp/client_r_1_c_1_req_1_resp_1_unary_1522345524.mem
Server CPU utilization: 17.173763239s
Server CPU profile: /tmp/Server_r_1_c_1_req_1_resp_1_unary_1522345524.cpu
Server Mem Profile: /tmp/Server_r_1_c_1_req_1_resp_1_unary_1522345524.mem
================================================================================
r_1_c_1_req_1_resp_1_streaming_1522345545
qps: 14345.3
Latency: (50/90/99 %ile): 67.821µs/83.16µs/104.562µs
Client CPU utilization: 22.728285653s
Client CPU profile: /tmp/client_r_1_c_1_req_1_resp_1_streaming_1522345545.cpu
Client Mem Profile: /tmp/client_r_1_c_1_req_1_resp_1_streaming_1522345545.mem
Server CPU utilization: 22.438695158s
Server CPU profile: /tmp/Server_r_1_c_1_req_1_resp_1_streaming_1522345545.cpu
Server Mem Profile: /tmp/Server_r_1_c_1_req_1_resp_1_streaming_1522345545.mem
================================================================================
r_1_c_1_req_100_resp_1_unary_1522345565
qps: 9155.9
Latency: (50/90/99 %ile): 102.661µs/125.137µs/176.439µs
Client CPU utilization: 15.744705954s
Client CPU profile: /tmp/client_r_1_c_1_req_100_resp_1_unary_1522345565.cpu
Client Mem Profile: /tmp/client_r_1_c_1_req_100_resp_1_unary_1522345565.mem
Server CPU utilization: 17.443043385s
Server CPU profile: /tmp/Server_r_1_c_1_req_100_resp_1_unary_1522345565.cpu
Server Mem Profile: /tmp/Server_r_1_c_1_req_100_resp_1_unary_1522345565.mem
================================================================================
r_1_c_1_req_100_resp_1_streaming_1522345587
qps: 13295.9
Latency: (50/90/99 %ile): 71.699µs/92.502µs/124.671µs
Client CPU utilization: 22.743129754s
Client CPU profile: /tmp/client_r_1_c_1_req_100_resp_1_streaming_1522345587.cpu
Client Mem Profile: /tmp/client_r_1_c_1_req_100_resp_1_streaming_1522345587.mem
Server CPU utilization: 22.615264368s
Server CPU profile: /tmp/Server_r_1_c_1_req_100_resp_1_streaming_1522345587.cpu
Server Mem Profile: /tmp/Server_r_1_c_1_req_100_resp_1_streaming_1522345587.mem
================================================================================
r_1_c_200_req_1_resp_1_unary_1522345607
qps: 81443.6
Latency: (50/90/99 %ile): 2.200862ms/3.746498ms/9.13501ms
Client CPU utilization: 1m40.708680062s
Client CPU profile: /tmp/client_r_1_c_200_req_1_resp_1_unary_1522345607.cpu
Client Mem Profile: /tmp/client_r_1_c_200_req_1_resp_1_unary_1522345607.mem
Server CPU utilization: 1m58.195342807s
Server CPU profile: /tmp/Server_r_1_c_200_req_1_resp_1_unary_1522345607.cpu
Server Mem Profile: /tmp/Server_r_1_c_200_req_1_resp_1_unary_1522345607.mem
================================================================================
r_1_c_200_req_1_resp_1_streaming_1522345628
qps: 149939.6
Latency: (50/90/99 %ile): 1.269794ms/1.962588ms/3.759668ms
Client CPU utilization: 1m49.214788204s
Client CPU profile: /tmp/client_r_1_c_200_req_1_resp_1_streaming_1522345628.cpu
Client Mem Profile: /tmp/client_r_1_c_200_req_1_resp_1_streaming_1522345628.mem
Server CPU utilization: 1m45.088960997s
Server CPU profile: /tmp/Server_r_1_c_200_req_1_resp_1_streaming_1522345628.cpu
Server Mem Profile: /tmp/Server_r_1_c_200_req_1_resp_1_streaming_1522345628.mem
================================================================================
r_1_c_200_req_100_resp_1_unary_1522345649
qps: 75291.4
Latency: (50/90/99 %ile): 2.332258ms/4.208357ms/9.884555ms
Client CPU utilization: 1m38.372587579s
Client CPU profile: /tmp/client_r_1_c_200_req_100_resp_1_unary_1522345649.cpu
Client Mem Profile: /tmp/client_r_1_c_200_req_100_resp_1_unary_1522345649.mem
Server CPU utilization: 1m54.293421801s
Server CPU profile: /tmp/Server_r_1_c_200_req_100_resp_1_unary_1522345649.cpu
Server Mem Profile: /tmp/Server_r_1_c_200_req_100_resp_1_unary_1522345649.mem
================================================================================
r_1_c_200_req_100_resp_1_streaming_1522345669
qps: 146473.7
Latency: (50/90/99 %ile): 1.286938ms/2.019942ms/4.119743ms
Client CPU utilization: 1m48.188005074s
Client CPU profile: /tmp/client_r_1_c_200_req_100_resp_1_streaming_1522345669.cpu
Client Mem Profile: /tmp/client_r_1_c_200_req_100_resp_1_streaming_1522345669.mem
Server CPU utilization: 1m43.985864847s
Server CPU profile: /tmp/Server_r_1_c_200_req_100_resp_1_streaming_1522345669.cpu
Server Mem Profile: /tmp/Server_r_1_c_200_req_100_resp_1_streaming_1522345669.mem
================================================================================
r_10_c_1_req_1_resp_1_unary_1522345690
qps: 41713
Latency: (50/90/99 %ile): 224.27µs/320.835µs/551.824µs
Client CPU utilization: 48.104101429s
Client CPU profile: /tmp/client_r_10_c_1_req_1_resp_1_unary_1522345690.cpu
Client Mem Profile: /tmp/client_r_10_c_1_req_1_resp_1_unary_1522345690.mem
Server CPU utilization: 54.867285235s
Server CPU profile: /tmp/Server_r_10_c_1_req_1_resp_1_unary_1522345690.cpu
Server Mem Profile: /tmp/Server_r_10_c_1_req_1_resp_1_unary_1522345690.mem
================================================================================
r_10_c_1_req_1_resp_1_streaming_1522345711
qps: 86132.1
Latency: (50/90/99 %ile): 109.487µs/151.798µs/235.289µs
Client CPU utilization: 49.221682659s
Client CPU profile: /tmp/client_r_10_c_1_req_1_resp_1_streaming_1522345711.cpu
Client Mem Profile: /tmp/client_r_10_c_1_req_1_resp_1_streaming_1522345711.mem
Server CPU utilization: 46.796757995s
Server CPU profile: /tmp/Server_r_10_c_1_req_1_resp_1_streaming_1522345711.cpu
Server Mem Profile: /tmp/Server_r_10_c_1_req_1_resp_1_streaming_1522345711.mem
================================================================================
r_10_c_1_req_100_resp_1_unary_1522345731
qps: 40753.6
Latency: (50/90/99 %ile): 227.362µs/331.936µs/583.229µs
Client CPU utilization: 47.645275133s
Client CPU profile: /tmp/client_r_10_c_1_req_100_resp_1_unary_1522345731.cpu
Client Mem Profile: /tmp/client_r_10_c_1_req_100_resp_1_unary_1522345731.mem
Server CPU utilization: 54.444179752s
Server CPU profile: /tmp/Server_r_10_c_1_req_100_resp_1_unary_1522345731.cpu
Server Mem Profile: /tmp/Server_r_10_c_1_req_100_resp_1_unary_1522345731.mem
================================================================================
r_10_c_1_req_100_resp_1_streaming_1522345751
qps: 85281
Latency: (50/90/99 %ile): 110.556µs/152.502µs/241.895µs
Client CPU utilization: 49.895785026s
Client CPU profile: /tmp/client_r_10_c_1_req_100_resp_1_streaming_1522345751.cpu
Client Mem Profile: /tmp/client_r_10_c_1_req_100_resp_1_streaming_1522345751.mem
Server CPU utilization: 47.539538087s
Server CPU profile: /tmp/Server_r_10_c_1_req_100_resp_1_streaming_1522345751.cpu
Server Mem Profile: /tmp/Server_r_10_c_1_req_100_resp_1_streaming_1522345751.mem
================================================================================
r_10_c_200_req_1_resp_1_unary_1522345772
qps: 103747
Latency: (50/90/99 %ile): 17.245362ms/32.398259ms/56.972638ms
Client CPU utilization: 1m42.37146451s
Client CPU profile: /tmp/client_r_10_c_200_req_1_resp_1_unary_1522345772.cpu
Client Mem Profile: /tmp/client_r_10_c_200_req_1_resp_1_unary_1522345772.mem
Server CPU utilization: 1m58.881603944s
Server CPU profile: /tmp/Server_r_10_c_200_req_1_resp_1_unary_1522345772.cpu
Server Mem Profile: /tmp/Server_r_10_c_200_req_1_resp_1_unary_1522345772.mem
================================================================================
r_10_c_200_req_1_resp_1_streaming_1522345793
qps: 216072.9
Latency: (50/90/99 %ile): 8.131356ms/15.229851ms/33.159954ms
Client CPU utilization: 1m56.269958794s
Client CPU profile: /tmp/client_r_10_c_200_req_1_resp_1_streaming_1522345793.cpu
Client Mem Profile: /tmp/client_r_10_c_200_req_1_resp_1_streaming_1522345793.mem
Server CPU utilization: 1m49.687516735s
Server CPU profile: /tmp/Server_r_10_c_200_req_1_resp_1_streaming_1522345793.cpu
Server Mem Profile: /tmp/Server_r_10_c_200_req_1_resp_1_streaming_1522345793.mem
================================================================================
r_10_c_200_req_100_resp_1_unary_1522345813
qps: 100977.6
Latency: (50/90/99 %ile): 17.829843ms/33.394985ms/56.336764ms
Client CPU utilization: 1m41.961430197s
Client CPU profile: /tmp/client_r_10_c_200_req_100_resp_1_unary_1522345813.cpu
Client Mem Profile: /tmp/client_r_10_c_200_req_100_resp_1_unary_1522345813.mem
Server CPU utilization: 1m57.15810443s
Server CPU profile: /tmp/Server_r_10_c_200_req_100_resp_1_unary_1522345813.cpu
Server Mem Profile: /tmp/Server_r_10_c_200_req_100_resp_1_unary_1522345813.mem
================================================================================
r_10_c_200_req_100_resp_1_streaming_1522345834
qps: 200728.4
Latency: (50/90/99 %ile): 8.647847ms/16.854851ms/35.64926ms
Client CPU utilization: 1m51.93917114s
Client CPU profile: /tmp/client_r_10_c_200_req_100_resp_1_streaming_1522345834.cpu
Client Mem Profile: /tmp/client_r_10_c_200_req_100_resp_1_streaming_1522345834.mem
Server CPU utilization: 1m45.537292628s
Server CPU profile: /tmp/Server_r_10_c_200_req_100_resp_1_streaming_1522345834.cpu
Server Mem Profile: /tmp/Server_r_10_c_200_req_100_resp_1_streaming_1522345834.mem
```

### CODE REVIEW

File: benchmark/run_bench.sh [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+      ;;
+    *)
+      echo "Incorrect option"
+      exit 1
+      ;;
+  esac
+done
+
+# Build server and client
+go install $GOPATH/src/google.golang.org/grpc/benchmark/server/benchserver.go && go install $GOPATH/src/google.golang.org/grpc/benchmark/client/benchclient.go
```
Comment (dfawley) [inline benchmark/run_bench.sh]: Is it possible to do this with `-o benchserver` and avoid renaming the files?

File: benchmark/client/main.go [hunk 2]
```diff
@@ -36,145 +37,160 @@ import (
 )
 
 var (
-	server            = flag.String("server", "", "The server address")
-	maxConcurrentRPCs = flag.Int("max_concurrent_rpcs", 1, "The max number of concurrent RPCs")
-	duration          = flag.Int("duration", math.MaxInt32, "The duration in seconds to run the benchmark client")
-	trace             = flag.Bool("trace", true, "Whether tracing is on")
-	rpcType           = flag.Int("rpc_type", 0,
+	port    = flag.String("port", "50051", "Localhost port to connect to.")
+	r       = flag.Int("r", 1, "The number of concurrent RPCs on each connection.")
```
Comment (dfawley) [inline benchmark/client/main.go]: These guys are global variables, and as such, should have long names.  Rationale:

1. Used far away from where defined.
2. Shadowed by commonly-used locals.

File: benchmark/run_bench.sh [hunk 3]
```diff
@@ -0,0 +1,207 @@
+#!/bin/bash
+
+rs=(1)
```
Comment (dfawley) [inline benchmark/run_bench.sh]: Same comment about names -- for readability, these should be descriptive.

File: benchmark/run_bench.sh [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+cs=(1)
+w=10
+d=10
+reqs=(1)
+resps=(1)
+rpc_types=(unary)
+
+# idx[0] = idx value for rs
```
Comment (dfawley) [inline benchmark/run_bench.sh]: I don't get it...what is idx?

File: benchmark/run_bench.sh [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  # Check to see if we have looped back to the beginning.
+  for v in ${idx[@]}; do
+    if [ ${v} == 0 ]; then
+      return=$((${return}*1))
+    else
+      return=$((${return}*0))
+    fi
+  done
+  if [ ${return} == 1 ]; then
+  rm -Rf ${out_dir}
```
Comment (dfawley) [inline benchmark/run_bench.sh]: indent please

File: benchmark/run_bench.sh [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+# idx[1] = idx value for cs
+# idx[2] = idx value for reqs
+# idx[3] = idx value for resps
+# idx[4] = idx value for rpc_types
+idx=(0 0 0 0 0)
+idx_max=(1 1 1 1 1)
+
+inc()
+{
+  for i in $(seq $((${#idx[@]}-1)) -1 0)
```
Comment (dfawley) [inline benchmark/run_bench.sh]: I prefer:

`for _____; do`

which you use below.  (Let's be consistent at the very least.)

File: benchmark/run_bench.sh [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  do
+    idx[${i}]=$((${idx[${i}]}+1))
+    if [ ${idx[${i}]} == ${idx_max[${i}]} ]; then
+      idx[${i}]=0
+    else
+      break
+    fi
+  done
+  return=1
```
Comment (dfawley) [inline benchmark/run_bench.sh]:
```
local return
return=1
```

File: benchmark/run_bench.sh [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+  # Check to see if we have looped back to the beginning.
+  for v in ${idx[@]}; do
+    if [ ${v} == 0 ]; then
+      return=$((${return}*1))
+    else
+      return=$((${return}*0))
```
Comment (dfawley) [inline benchmark/run_bench.sh]: return=0?

File: benchmark/run_bench.sh [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    if [ ${idx[${i}]} == ${idx_max[${i}]} ]; then
+      idx[${i}]=0
+    else
+      break
+    fi
+  done
+  return=1
+  # Check to see if we have looped back to the beginning.
+  for v in ${idx[@]}; do
+    if [ ${v} == 0 ]; then
```
Comment (dfawley) [inline benchmark/run_bench.sh]: Should this just be `if [ $v != 0 ]; then return=0; fi`

File: benchmark/run_bench.sh (hunk 7 above)
Comment (dfawley) [inline benchmark/run_bench.sh]: `return` is a bash keyword; let's not use it for a variable name, please.

File: benchmark/run_bench.sh [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+      return=$((${return}*1))
+    else
+      return=$((${return}*0))
+    fi
+  done
+  if [ ${return} == 1 ]; then
+  rm -Rf ${out_dir}
+    exit 0
```
Comment (dfawley) [inline benchmark/run_bench.sh]: `exit` terminates the script; did you want `return`?

File: benchmark/run_bench.sh [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+  fi
+}
+
+run(){
+  nr=${rs[${idx[0]}]}
```
Comment (dfawley) [inline benchmark/run_bench.sh]: needs better names.  Also should be local (since in a function), as above.

File: benchmark/run_bench.sh [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+    if [ ${delta} == 10 ]; then
+      echo "Continuous 10 failed runs. Exiting now."
+      rm -Rf ${out_dir}
+      exit 1
+    fi
+  done
+
+}
+
+while test $# -gt 0; do
```
Comment (dfawley) [inline benchmark/run_bench.sh]: Use `[ $# -gt 0 ]` instead for consistency with code above or `[[ $# > 0 ]]` for efficiency/clarity, but non-sh-compliance.

File: benchmark/run_bench.sh [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+  case "$1" in
+    -r)
+      shift
+      if test $# -gt 0; then
```
Comment (dfawley) [inline benchmark/run_bench.sh]: prefer:

```
if [ $# -eq 0 ]; then   # or [[ $# == 0 ]]
  echo "error stuff"
  exit 1
fi
rs=....
...
```

File: benchmark/run_bench.sh [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+      shift
+      ;;
+    -rpc_type)
+      shift
+      if test $# -gt 0; then
+        rpc_types=($(echo $1 | sed 's/,/ /g'))
+        idx_max[4]=${#rpc_types[@]}
+        for val in ${rpc_types[@]}; do
+          case "${val}" in
+            "unary"|"streaming");;
```
Comment (dfawley) [inline benchmark/run_bench.sh]: do we need to vet the vals here, or can we pass them to the script and wait for it to fail?

File: benchmark/run_bench.sh [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+        done
+      else
+        echo "rpc type not specified not specified"
+        exit 1
+      fi
+      shift
+      ;;
+    -h|--help)
+      echo "Following are valid options:"
+      echo ""
```
Comment (dfawley) [inline benchmark/run_bench.sh]: `echo ""` == `echo`

File: benchmark/run_bench.sh [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+        idx_max[0]=${#rs[@]}
+      else
+        echo "number of rpcs not specified"
+        exit 1
+      fi
+      shift
+      ;;
+    -c)
+      shift
+      if test $# -gt 0; then
```
Comment (dfawley) [inline benchmark/run_bench.sh]: Consider a refactor:

```
param() {
  argname=$1
  shift
  if [ $# -eq 0 ]; then
    echo "$argname not specified"
    exit 1
  fi
  echo $1 | sed 's/,/ /g'
}
...
  -c)
    shift
    cs=$(param "number of connections" "$@")
    shift
```

File: benchmark/run_bench.sh [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+  done
+
+}
+
+while test $# -gt 0; do
+  case "$1" in
+    -r)
+      shift
+      if test $# -gt 0; then
+        rs=($(echo $1 | sed 's/,/ /g'))
```
Comment (dfawley) [inline benchmark/run_bench.sh]: why a subshell?

This is preferred: `rs="$(echo "$1" | sed 's/,/ /g')"`

File: benchmark/run_bench.sh (hunk 4 above)
Comment (MakMukhi) [reply benchmark/run_bench.sh]: All input parameters that are arrays, need an index to go over them while keeping other parameters' index unchanged. This way we can generate all combinations without using nested for loops

File: benchmark/run_bench.sh (hunk 16 above)
Comment (MakMukhi) [reply benchmark/run_bench.sh]: `cs=$(param "arg1" "arg2")` doesn't call `param()`. Maybe, I'm doing something wrong.

File: benchmark/run_bench.sh (hunk 10 above)
Comment (MakMukhi) [reply benchmark/run_bench.sh]: `inc` is called from an infinite loop. `exit 0` makes sure that we stop the script when we are done with all combinations.

File: benchmark/run_bench.sh (hunk 17 above)
Comment (dfawley) [reply benchmark/run_bench.sh]: Oh yes, I see...parens are overloaded.

File: benchmark/run_bench.sh (hunk 16 above)
Comment (dfawley) [reply benchmark/run_bench.sh]: Maybe so?

```
$ x() { echo "hi"; }
$ y=$(x)
$ echo $y
hi
```

File: benchmark/run_bench.sh [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+  if [ ${idx} -lt 0 ]; then
+    return
+  fi
+  idx_max[${idx}]=${#PARAM[@]}
+}
+
+while [ $# -gt 0 ]; do
+  case "$1" in
+    -r)
+      shift
```
Comment (dfawley) [inline benchmark/run_bench.sh]: You can (re)move these shifts:

```
while ... do
  case "$1" in
    -r)
      set_param "number of rpcs" 0 $2
      rpcs=...
      ;;
...
  esac
  shift; shift
done
```

**Pull request merged**